src/algebra/graded_monoid.lean

```diff
@@ -448,10 +488,37 @@ instance set_like.ghas_mul {S : Type*} [set_like S R] [has_mul R] [has_add ι] (
   [set_like.has_graded_mul A] {i j : ι} (x : A i) (y : A j) :
     ↑(@graded_monoid.ghas_mul.mul _ (λ i, A i) _ _ _ _ x y) = (x * y : R) := rfl
 
+/-- A version of `graded_monoid.ghas_smul` for internally graded objects. -/
+class set_like.has_graded_smul {S R N M : Type*} [set_like S R] [set_like N M]
+  [has_smul R M] [has_add ι] (A : ι → S) (B : ι → N) : Prop :=
+(smul_mem : ∀ ⦃i j : ι⦄ {ai bj}, ai ∈ A i → bj ∈ B j → ai • bj ∈ B (i + j))
+
+lemma set_like.smul_mem_graded {S R N M : Type*} [set_like S R] [set_like N M]
+  [has_smul R M] [has_add ι] (A : ι → S) (B : ι → N) [set_like.has_graded_smul A B]
+  ⦃i j⦄ {gi gj} (hi : gi ∈ A i) (hj : gj ∈ B j) :
+  gi • gj ∈ B (i + j) :=
+set_like.has_graded_smul.smul_mem hi hj
+
+instance set_like.ghas_smul {S R N M : Type*} [set_like S R] [set_like N M]
+  [has_smul R M] [has_add ι] (A : ι → S) (B : ι → N) [set_like.has_graded_smul A B] :
+  graded_monoid.ghas_smul (λ i, A i) (λ i, B i) :=
+{ smul := λ i j a b, ⟨(a : R) • b, set_like.has_graded_smul.smul_mem a.2 b.2⟩ }
+
+@[simp] lemma set_like.coe_ghas_smul {S R N M : Type*} [set_like S R] [set_like N M]
+  [has_smul R M] [has_add ι] (A : ι → S) (B : ι → N) [set_like.has_graded_smul A B]
+  {i j : ι} (x : A i) (y : B j) :
+  (@graded_monoid.ghas_smul.smul ι (λ i, A i) (λ i, B i) _ _ i j x y : M) = ((x : R) • y) :=
+rfl
+
 /-- A version of `graded_monoid.gmonoid` for internally graded objects. -/
 class set_like.graded_monoid {S : Type*} [set_like S R] [monoid R] [add_monoid ι]
   (A : ι → S) extends set_like.has_graded_one A, set_like.has_graded_mul A : Prop
 
+instance set_like.graded_monoid.to_has_graded_smul
```

Comment:
This should only need `set_like.has_graded_mul`; there should be a stronger version that says `set_like.graded_monoid` implies `gmul_action`.

Reply:
Is this resolved now?